driver-core/src/main/com/mongodb/internal/connection/SingleServerCluster.java

```diff
@@ -52,38 +55,39 @@ public SingleServerCluster(final ClusterId clusterId, final ClusterSettings sett
             LOGGER.info(format("Cluster created with settings %s", settings.getShortDescription()));
         }
 
+        server = new AtomicReference<>();
         // synchronized in the constructor because the change listener is re-entrant to this instance.
         // In other words, we are leaking a reference to "this" from the constructor.
-        synchronized (this) {
-            this.server = createServer(settings.getHosts().get(0), new DefaultServerDescriptionChangedListener());
+        withLock(() -> {
+            server.set(createServer(settings.getHosts().get(0), new DefaultServerDescriptionChangedListener()));
             publishDescription(ServerDescription.builder().state(CONNECTING).address(settings.getHosts().get(0))
                     .build());
-        }
+        });
     }
 
     @Override
     protected void connect() {
-        server.connect();
+        assertNotNull(server.get()).connect();
     }
 
     @Override
     public ClusterableServer getServer(final ServerAddress serverAddress) {
         isTrue("open", !isClosed());
-        return server;
+        return assertNotNull(server.get());
     }
 
     @Override
     public void close() {
         if (!isClosed()) {
-            server.close();
+            assertNotNull(server.get()).close();
             super.close();
         }
     }
 
     private class DefaultServerDescriptionChangedListener implements ServerDescriptionChangedListener {
         @Override
         public void serverDescriptionChanged(final ServerDescriptionChangedEvent event) {
-            synchronized (this) {
+            withLock(() -> {
```

Comment:
Note that the monitor used here for synchronization (`DefaultServerDescriptionChangedListener.this`) was different than the one used as a result of changes in this PR (`SingleServerCluster.this`). However, I don't think there was a reason to use a different one. Note that `AbstractMultiServerCluster` used and continues to use the same monitor `AbstractMultiServerCluster.this` both to handle events in its `AbstractMultiServerCluster.DefaultServerDescriptionChangedListener` and in other places.

Reply:
Ack